pkg/syncer/controllermanager/controllermanager.go

```diff
@@ -0,0 +1,193 @@
+	controllerManager := ControllerManager{
+		name:               ControllerNamePrefix + suffix,
+		queue:              workqueue.NewNamedRateLimitingQueue(workqueue.DefaultControllerRateLimiter(), ControllerNamePrefix+suffix),
+		informerSource:     informerSource,
+		managedControllers: controllers,
+		startedControllers: map[string]context.CancelFunc{},
+	}
+
+	apisChanged := informerSource.Subscribe(controllerManager.name)
+
+	logger := klog.FromContext(ctx)
+
+	go func() {
+		for {
+			select {
+			case <-ctx.Done():
+				return
+			case <-apisChanged:
+				logger.V(4).Info("got API change notification")
+				controllerManager.queue.Add("resync") // this queue only ever has one key in it, as long as it's constant we are OK
+			}
+		}
+	}()
+
+	return &controllerManager
+}
+
+// ControllerManager is a component that manages (create/start/stop) GVR-specific controllers according to available GVRs.
+// It reacts to the changes of supported GVRs in a DiscoveringDynamicSharedInformerFactory
+// (the GVRs for which an informer has been automatically created, started and synced),
+// and starts / stops registered GVRs-specific controllers according to the GVRs they depend on.
+//
+// For example this allows starting PVC / PV controllers only when PVC / PV resources are exposed by the Syncer and UpSyncer
+// virtual workspaces, and Informers for them have been started and synced by the corresponding ddsif.
+type ControllerManager struct {
+	name               string
+	queue              workqueue.RateLimitingInterface
+	informerSource     InformerSource
+	managedControllers map[string]ManagedController
+	startedControllers map[string]context.CancelFunc
+}
+
+// Start starts the controller, which stops when ctx.Done() is closed.
+func (c *ControllerManager) Start(ctx context.Context) {
+	defer utilruntime.HandleCrash()
+	defer c.queue.ShutDown()
+
+	logger := logging.WithReconciler(klog.FromContext(ctx), c.name)
+	logger.Info("Starting controller manager")
+	defer logger.Info("Shutting down controller manager")
+
+	go wait.UntilWithContext(ctx, c.startWorker, time.Second)
+	<-ctx.Done()
+}
+
+func (c *ControllerManager) startWorker(ctx context.Context) {
+	for c.processNextWorkItem(ctx) {
+	}
+}
+
+func (c *ControllerManager) processNextWorkItem(ctx context.Context) bool {
+	key, quit := c.queue.Get()
+	if quit {
+		return false
+	}
+	defer c.queue.Done(key)
+
+	c.updateControllers(ctx)
+	c.queue.Forget(key)
+	return true
+}
+
+func (c *ControllerManager) updateControllers(ctx context.Context) {
+	logger := klog.FromContext(ctx)
+	controllersToStart := map[string]CreateControllerFunc{}
+controllerLoop:
+	for controllerName, managedController := range c.managedControllers {
+		requiredGVRs := managedController.RequiredGVRs
+		for _, gvr := range requiredGVRs {
+			if informer, known := c.informerSource.Informer(gvr); !known {
+				continue controllerLoop
```

Comment:
I don't understand this logic. Why should we not wait for sync on known==false ?

Reply:
because if the informer doesn't exist for this GVR, and would later be added in the informer source, then the InformerSource (ddsif) will notify the controller manager that supported GVRs changed, and the informer will be known this time.

Reply:
nvmd. Misunderstood known. `exists` would be a better word.

Reply:
> `exists` would be a better word.

You're right. I'll change this.

Reply:
Fixed in commit da89d306186c852a5f9f9ab2a3a549f896f5f7e0